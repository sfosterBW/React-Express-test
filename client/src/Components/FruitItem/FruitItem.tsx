import React, { FC, useState } from 'react'
import { useDispatch } from 'react-redux'
import { Fruit } from '../../utils/interfaces'
import { openModal } from '../../Reducers/modalReducer'
import { removeFruit, updateFruit } from '../../Reducers/fruitReducer'
import styles from './FruitItem.module.css'

interface Props {
  active?: boolean
  fruit: Fruit
}

const FruitItem: FC<Props> = ({ active = false, fruit }) => {
  const dispatch = useDispatch()

  const [activeToggle, setActiveToggle] = useState<boolean>(active)
  
  const { id, name, best, description } = fruit

  return (
    <div className={styles.row} data-testid="item">
      <div
        className={styles.header}
        onClick={() => setActiveToggle(!activeToggle)}
      >
        <div className={styles.section}>
          <h3 className={styles.subtitle} data-testid="item-title">
            {name}
          </h3>
        </div>
      </div>
      <div
        className={styles.active}
        style={{
          height: activeToggle ? "calc(120px + 12vmin)" : "0",
          transition: "all 0.5s"
        }}
      >
        <div className={styles.sectionRow}>
          <h4 className={styles.subtitle}>
            Description
          </h4>
          <p className={styles.description} data-testid="item-description">
            {description ? `${description}` : "No description added yet"}
          </p>
        </div>
        <div className={styles.section}>
          <h4 className={styles.subtitle}>
            Best
          </h4>
          <input
            checked={best}
            className={styles.button}
            id={id}
            name={id}
            onChange={() => {
              dispatch(updateFruit({ ...fruit, best: !best }))
            }}
            type="checkbox"
            value={fruit.id}
          />
          <label data-testid="item-best" htmlFor={id}>
            {`${best}`}
          </label>
        </div>
        <div className={styles.section}>
          <h4 className={styles.subtitle}>
            Edit
          </h4>
          <button
            className={styles.button}
            data-testid="update-button"
            name="modal"
            onClick={() => dispatch(openModal(fruit))}
          >
            Update
          </button>
          <button
            className={styles.button}
            data-testid="remove-button"
            name="remove"
            onClick={() => dispatch(removeFruit(id))}
            value={`${id}`}
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  )
}

export default FruitItem
