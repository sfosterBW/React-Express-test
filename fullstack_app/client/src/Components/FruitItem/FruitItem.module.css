.row {
  border-radius: 4px;
  width: calc(120px + 20vmin);
  margin: 1vmin 0;
}

.section {
  display: flex;
  box-sizing: border-box;
  align-items: center;
  justify-content: center;
  padding: 2vmin;
  width: 100%;
}

.header {
  padding: 1vmin;
  background: #1ab545cc;
  color: white;
  text-shadow: 0.2vmin 0.2vmin gray;
  font-size: calc(12px + 2vmin);
}

.header:hover {
  box-shadow: 0.2vmin 0.2vmin 0.5vmin 0.5vmin rgba(0, 0, 0, 0.25);
  cursor: pointer;
}

.active {
  display: flex;
  box-sizing: border-box;
  align-items: center;
  flex-direction: column;
  overflow: hidden;
}

.subtitle {
  margin: 0 2vmin;
  text-align: center;
  font-size: calc(10px + 2vmin);
}

.button {
  margin-right: 1vmin;
}
