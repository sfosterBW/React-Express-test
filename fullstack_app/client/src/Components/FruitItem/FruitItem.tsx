import React, { FC, useState } from 'react'
import { useDispatch } from 'react-redux'
import { IFruit } from '../../utils/interfaces'
import fruitService from '../../utils/api'
import { openModal, removeFruit, updateFruit } from '../../utils/actions'
import styles from './FruitItem.module.css'

interface Props {
  active?: boolean
  fruit: IFruit
}

const FruitItem: FC<Props> = ({ active = false, fruit }) => {
  const [activeToggle, setActiveToggle] = useState<boolean>(active)
  const dispatch = useDispatch()

  const handleEdit = async (fruit: IFruit) => {
    const newFruit = fruit
    newFruit.best = !newFruit.best

    const res = await fruitService.updateFruit(newFruit)
    dispatch(updateFruit(res.data))
  }

  const handleRemove = async (fruit: IFruit) => {
    if(fruit.id){
      const res = await fruitService.deleteFruit(fruit.id)
      dispatch(removeFruit(String(res.data)))
    } else {
      console.log("Fruit undefined", fruit)
    }
  }

  const showHide = () => activeToggle ? "calc(60px + 8vmin)" : "0"

  return (
    <div className={styles.row}>
      <div
        className={styles.header}
        onClick={() => setActiveToggle(!activeToggle)}
      >
        <div className={styles.section}>
          <h3 className={styles.subtitle}>
            {fruit.name}
          </h3>
        </div>
      </div>
      <div
        className={styles.active}
        style={{ height: showHide(), transition: "all 0.5s" }}
      >
        <div className={styles.section}>
          <h3 className={styles.subtitle}>
            Best
          </h3>
          <input
            checked={fruit.best}
            className={styles.button}
            id={`${fruit.id}`}
            name={`${fruit.id}`}
            onChange={() => handleEdit(fruit)}
            type="checkbox"
            value={fruit.id}
          />
          <label htmlFor={`${fruit.id}`}>
            {fruit.best.toString()}
          </label>
        </div>
        <div className={styles.section}>
          <h3 className={styles.subtitle}>
            Edit
          </h3>
          <button
            className={styles.button}
            name="modal"
            onClick={() => dispatch(openModal(fruit))}
          >
            Update
          </button>
          <button
            className={styles.button}
            name="remove"
            onClick={() => handleRemove(fruit)}
            value={`${fruit.id}`}
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  )
}

export default FruitItem
