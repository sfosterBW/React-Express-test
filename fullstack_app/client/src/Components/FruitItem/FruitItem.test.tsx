
import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import FruitItem from './FruitItem'

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}))

describe('the FruitItem component', () => {

  const fruit = {id: "1", name:"apple", best:false}
  const component = <FruitItem fruit={fruit} />
  const wrapper = mount(component)

  it('renders with the correct structure', () => {
    expect(wrapper).toBeDefined()
    expect(wrapper.find('div')).toHaveLength(6)
    expect(wrapper.find('h3')).toHaveLength(3)
    expect(wrapper.find('h3').at(0).text()).toEqual(fruit.name)
    expect(wrapper.find('h3').at(1).text()).toEqual("Best")
    expect(wrapper.find('label')).toHaveLength(1)
    expect(wrapper.find('label').text()).toEqual(String(fruit.best))
    expect(wrapper.find('label').props().htmlFor).toEqual(String(fruit.id))
    expect(wrapper.find('input')).toHaveLength(1)
    expect(wrapper.find('input').props().name).toEqual(String(fruit.id))
    expect(wrapper.find('input').props().type).toEqual('checkbox')
    expect(wrapper.find('button')).toHaveLength(2)
    expect(wrapper.find('button').at(0).props().name).toEqual('modal')
    expect(wrapper.find('button').at(1).props().name).toEqual('remove')
  })

  it('functions as expected',() => {
    wrapper.find('button').at(0).simulate('click')
    wrapper.find('button').at(1).simulate('click')
    wrapper.find('input').simulate('change')
    expect(mockDispatch).toHaveBeenCalledTimes(1)
  })

  it('renders the same as last time', () => {
    const tree = renderer
      .create(component)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
