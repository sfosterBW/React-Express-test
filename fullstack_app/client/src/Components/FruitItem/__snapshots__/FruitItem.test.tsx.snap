// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the FruitItem component renders the same as last time 1`] = `
<div
  className="row"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="section"
    >
      <h3
        className="subtitle"
      >
        apple
      </h3>
    </div>
  </div>
  <div
    className="active"
    style={
      Object {
        "height": "0",
        "transition": "all 0.5s",
      }
    }
  >
    <div
      className="section"
    >
      <h3
        className="subtitle"
      >
        Best
      </h3>
      <input
        checked={true}
        className="button"
        id="1"
        name="1"
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        htmlFor="1"
      >
        true
      </label>
    </div>
    <div
      className="section"
    >
      <h3
        className="subtitle"
      >
        Edit
      </h3>
      <button
        className="button"
        name="modal"
        onClick={[Function]}
      >
        Update
      </button>
      <button
        className="button"
        name="remove"
        onClick={[Function]}
        value="1"
      >
        Remove
      </button>
    </div>
  </div>
</div>
`;
